GitLab: Add support to use threads instead of comments

uncaught opened the pull request:
- Add support to use threads instead of comments
- Implements #1138 

# Implementation

- Threads are used if an environment variable `DANGER_GITLAB_USE_THREADS` is set to a value of `1` or `true`.
- Threads are completely removed with all answers (even of other users) if danger removes its main comment.

## GitLab API change

- A new method `getMergeRequestDiscussions` fetches all the threads (discussions) of a merge request
- **BREAKING** The method `createMergeRequestDiscussion` has had its second argument changed from being the position object to be the actual options object for that gitbreaker API call, which contains the position.

# Migration considerations

- You can switch to using threads and back any time.
- Only newly created comments are affected by the current setting.
- Because existing comments can be turned into a thread by simply replying to them, danger will always delete its entire threads, no matter the setting.

orta commented:
Hey, this looks  good to me from a read over it - I'm quite wary of this change being so substantial and only really having one test to ensure it doesn't break in the future. So, I'll give it a week to settle, _if you can_ it'd be good to get some tests - if not, that's fine.

uncaught commented:
Thanks, @orta, I've added some tests - there were none at all before for the GitLab.ts.

I've also changed one key detail: danger will always delete its entire threads now, no matter the setting. Because I've noticed that you can simply turn danger's comments into threads by replying to them as another user. And that would leave the replies dangling if danger was then to remove its original comment.

[GitLab's discussions API](https://docs.gitlab.com/ee/api/discussions.html#list-project-issue-discussion-items) doesn't distinguish between comments and threads. You will always get all comments, wrapped in a discussion object from that API.

orta commented:
Great, this looks good to me - I've given it a once over and will give it a second look and then merge soon!

orta reviewed (approved):
Nice!

orta commented:
Thanks for the substantial PR! This has shipped as danger 11.2.0

squarefrog commented:
This PR has broken thread reporting for us, using GitLab 15.3. Using danger 11.1.4 we still get both inline and global warnings.

Using danger 11.2.0 or later, we only get global warnings, as they replace the body of the inline warnings. Full context available in https://github.com/danger/danger-js/issues/1351

samitha9125 commented:
I second what @squarefrog  says. New feature delete old comments/threads in a MR as well when enabled.
